### PR TITLE
feat: Add new route for issuing OAuth2 access_token

### PR DESCRIPTION
**Pull request by h-kanazawa**
https://ht2labs.myjetbrains.com/youtrack/issue/LL-698

### CODE REVIEW

File: api/src/server.js [hunk 1]
```diff
@@ -18,6 +18,7 @@ app.use(corsMiddleware);
 app.options('*', corsMiddleware);
 
 app.use(bodyParser.json({ limit: '5mb' }));
+app.use(bodyParser.urlencoded({ extended: false }));
```
Comment (ryansmith94) [inline api/src/server.js]: @ht2 do you foresee any issues with this?

File: cli/src/commands/v2-migrations/20190131000000_oauth_tokens.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { getConnection } from 'lib/connections/mongoose';
+
+const up = async () => {
+  const connection = getConnection();
+  await connection.collection('oAuthTokens').createIndex({ expireAt: 1 }, { expireAfterSeconds: 0 });
```
Comment (ryansmith94) [inline cli/src/commands/v2-migrations/20190131000000_oauth_tokens.js]: Nice job, glad you were able to figure this out 👍

File: api/src/server.js (hunk 1 above)
Comment (h-kanazawa) [reply api/src/server.js]: @ht2 for your reference
https://github.com/expressjs/body-parser#bodyparserurlencodedoptions

> Returns middleware that only parses urlencoded bodies and **only looks at requests where the Content-Type header matches the type option.**

> Options
> ...
> **type**
> ... Defaults to `application/x-www-form-urlencoded`

File: api/src/auth/passport.js [hunk 3]
```diff
@@ -224,4 +224,64 @@ if (
...
+      done({ error: 'unsupported_grant_type' });
+      return;
+    }
+
+    if (clientId === undefined || clientSecret === undefined) {
+      done({ error: 'invalid_request' });
+      return;
+    }
+
+    Client.findOne({ 'api.basic_key': clientId }, (err, client) => {
```
[the remark this answers is not in the document]
Comment (h-kanazawa) [reply api/src/auth/passport.js]: All right. I include `api.basic_key`, `api.basic_secret`, and `isTrusted` checking in the DB call. I remove the error description.

File: api/src/controllers/AuthController.js [hunk 4]
```diff
@@ -214,12 +220,52 @@ const success = (req, res) => {
   res.send('Login success!');
 };
 
+const issueOAuth2AccessToken = (req, res) => {
+  passport.authenticate('OAuth2_Authorization', DEFAULT_PASSPORT_OPTIONS, (err, client) => {
+    if (err) {
+      if (err.isServerError) {
```
[the remark this answers is not in the document]
Comment (h-kanazawa) [reply api/src/controllers/AuthController.js]: I agree. Default should be server error and Client error should be explicitly declared.

File: api/src/server.js (hunk 1 above)
Comment (ht2) [reply api/src/server.js]: Noted - no can't see any big issues here

Comment (ryansmith94) [reply api/src/server.js]: Ok great thanks for checking 👍

File: api/src/auth/passport.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryansmith94) [reply api/src/auth/passport.js]: Ok cool thanks Hota 👍

**Comment by ryansmith94** [comment]
Note for Asahd: Tested migration worked.

**Pull request merged**